packages/drivers/odsp-driver/src/fetchWithEpochValidation.ts

```diff
@@ -0,0 +1,125 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import assert from "assert";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { createOdspNetworkError, fluidEpochMismatchError, OdspErrorType } from "@fluidframework/odsp-doclib-utils";
+import { fetchAndParseAsJSONHelper, fetchHelper, IOdspResponse } from "./odspUtils";
+import { ICacheEntry, IPersistedCache, IPersistedCacheValue } from "./odspCache";
+
+/**
+ * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
+ * server can match it with its epoch value in order to match the version.
+ * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
+ * then it also clears all the cached entries for the given container.
+ */
+export class FetchWithEpochValidation {
+    private _fluidEpoch: string | undefined;
+    private _hashedDocumentId: string | undefined;
+    constructor(
+        private readonly persistedCache: IPersistedCache,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
+        this._hashedDocumentId = docId;
+    }
+
+    public get hashedDocumentId(): string | undefined {
+        return this._hashedDocumentId;
+    }
+
+    public get fluidEpoch() {
+        return this._fluidEpoch;
+    }
+
+    public async fetchFromCache<T>(entry: ICacheEntry, maxOpCount: number | undefined): Promise<T | undefined> {
+        const value: IPersistedCacheValue = await this.persistedCache.get(entry, maxOpCount);
+        if (value !== undefined) {
+            try {
+                this.validateEpochFromResponse(value.fluidEpoch);
+            } catch (error) {
+                this.checkForEpochError(createOdspNetworkError("Epoch Mismatch", fluidEpochMismatchError));
+            }
+            return value.value as T;
+        }
+        return undefined;
+    }
+
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
```

Comment:
Do we need this argument? What is the guidance to the caller when to set it to false or true?
I think we can simplify here and always make this choice here (not through default arg)

Reply:
This should be set to true if the caller wants to add it in the post body like for the POST call for trees latest. Will add the description. But is difficult and will be messy to determine here whether to add it in body or header because the body, headers etc are already prepared at this stage. Once we move to all post apis, we can remove this argument.